File: sdk/python/shark_auth/api_keys.py
"""Admin API-key management.

Wraps ``/api/v1/api-keys`` (mounted under the admin-key auth group; see
``internal/api/router.go``).
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from . import _http
from .proxy_rules import _raise


class APIKeysClient:
    """Admin client for managing admin API keys.

    Parameters
    ----------
    base_url:
        Base URL of the SharkAuth server.
    admin_api_key:
        Admin API key (``sk_live_...``).
    session:
        Optional pre-configured :class:`requests.Session`.
    """

    _PREFIX = "/api/v1/api-keys"

    def __init__(
        self,
        base_url: str,
        admin_api_key: str,
        *,
        session: Optional[object] = None,
    ) -> None:
        self._base = base_url.rstrip("/")
        self._token = admin_api_key
        self._session = session or _http.new_session()

    def _auth(self) -> Dict[str, str]:
        return {"Authorization": f"Bearer {self._token}"}

    @staticmethod
    def _unwrap(body: Any) -> Any:
        if isinstance(body, dict) and "data" in body and len(body) <= 2:
            return body["data"]
        return body

    def create(
        self,
        name: str,
        scopes: Optional[List[str]] = None,
        expires_at: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Create a new admin API key. POST /api/v1/api-keys.

        Returns a dict containing both ``key_id`` (or ``id``) *and* the
        full ``key`` value. The full key is only returned at creation —
        store it immediately.
        """
        body: Dict[str, Any] = {"name": name}
        if scopes is not None:
            body["scopes"] = scopes
        if expires_at is not None:
            body["expires_at"] = expires_at
        url = f"{self._base}{self._PREFIX}"
        resp = _http.request(self._session, "POST", url, headers=self._auth(), json=body)
        if resp.status_code in (200, 201):
            return self._unwrap(resp.json())
        _raise(resp)

    def list(self) -> List[Dict[str, Any]]:
        """GET /api/v1/api-keys."""
        url = f"{self._base}{self._PREFIX}"
        resp = _http.request(self._session, "GET", url, headers=self._auth())
        if resp.status_code == 200:
            body = resp.json()
            if isinstance(body, dict):
                return body.get("data", []) or body.get("api_keys", []) or []
            return body
        _raise(resp)

    def get(self, key_id: str) -> Dict[str, Any]:
        """GET /api/v1/api-keys/{id}."""
        url = f"{self._base}{self._PREFIX}/{key_id}"
        resp = _http.request(self._session, "GET", url, headers=self._auth())
        if resp.status_code == 200:
            return self._unwrap(resp.json())
        _raise(resp)

    def revoke(self, key_id: str) -> None:
        """Revoke (delete) an API key. DELETE /api/v1/api-keys/{id}."""
        url = f"{self._base}{self._PREFIX}/{key_id}"
        resp = _http.request(self._session, "DELETE", url, headers=self._auth())
        if resp.status_code in (200, 204):
            return None
        _raise(resp)

    def rotate(self, key_id: str) -> Dict[str, Any]:
        """Rotate an API key. POST /api/v1/api-keys/{id}/rotate.

        Returns a dict with the new ``key`` and (typically) a new
        ``key_id``. Store the new key immediately — the server cannot
        return it again.
        """
        url = f"{self._base}{self._PREFIX}/{key_id}/rotate"
        resp = _http.request(self._session, "POST", url, headers=self._auth())
        if resp.status_code in (200, 201):
            return self._unwrap(resp.json())
        _raise(resp)

File: sdk/typescript/src/apiKeys.ts
/**
 * Admin API-key management.
 *
 * Wraps `/api/v1/api-keys` (mounted under the admin-key auth group; see
 * `internal/api/router.go`).
 */

import { httpRequest } from "./http.js";
import { SharkAPIError } from "./errors.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** API key row as returned by the server. */
export interface ApiKey {
  id?: string;
  key_id?: string;
  name?: string;
  scopes?: string[];
  expires_at?: string;
  created_at?: string;
  /** Plaintext key — only returned at creation/rotation. */
  key?: string;
  [key: string]: unknown;
}

/** Input for {@link ApiKeysClient.create}. */
export interface CreateApiKeyInput {
  name: string;
  scopes?: string[];
  /** ISO 8601 expiry. */
  expires_at?: string;
}

/** Options for {@link ApiKeysClient}. */
export interface ApiKeysClientOptions {
  baseUrl: string;
  adminKey: string;
}

const PREFIX = "/api/v1/api-keys";

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

/** Admin client for managing admin API keys. */
export class ApiKeysClient {
  private readonly _base: string;
  private readonly _key: string;

  constructor(opts: ApiKeysClientOptions) {
    this._base = opts.baseUrl.replace(/\/+$/, "");
    this._key = opts.adminKey;
  }

  private _auth(): Record<string, string> {
    return { Authorization: `Bearer ${this._key}` };
  }

  private async _throw(status: number, text: string): Promise<never> {
    let code = "api_error";
    let message = text.slice(0, 300);
    try {
      const body = JSON.parse(text) as {
        error?: { code?: string; message?: string } | string;
        message?: string;
      };
      if (typeof body.error === "object" && body.error) {
        if (body.error.code) code = body.error.code;
        if (body.error.message) message = body.error.message;
      } else if (typeof body.error === "string") {
        message = body.error;
      } else if (body.message) {
        message = body.message;
      }
    } catch {
      // keep defaults
    }
    throw new SharkAPIError(message, code, status);
  }

  private static _unwrap<T>(body: unknown): T {
    if (
      body &&
      typeof body === "object" &&
      "data" in (body as Record<string, unknown>) &&
      Object.keys(body as Record<string, unknown>).length <= 2
    ) {
      return (body as { data: T }).data;
    }
    return body as T;
  }

  /**
   * Create a new admin API key.
   *
   * Returns a row containing both `key_id` (or `id`) AND the full `key`
   * value. The full key is only returned at creation — store it now.
   */
  async create(input: CreateApiKeyInput): Promise<ApiKey> {
    const body: Record<string, unknown> = { name: input.name };
    if (input.scopes !== undefined) body.scopes = input.scopes;
    if (input.expires_at !== undefined) body.expires_at = input.expires_at;
    const url = `${this._base}${PREFIX}`;
    const resp = await httpRequest(url, {
      method: "POST",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    if (resp.status !== 200 && resp.status !== 201)
      await this._throw(resp.status, resp.text);
    return ApiKeysClient._unwrap<ApiKey>(resp.json());
  }

  /** List API keys. */
  async list(): Promise<ApiKey[]> {
    const url = `${this._base}${PREFIX}`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    const body = resp.json<ApiKey[] | { data?: ApiKey[]; api_keys?: ApiKey[] }>();
    if (Array.isArray(body)) return body;
    return body?.data ?? body?.api_keys ?? [];
  }

  /** Fetch a single API key. */
  async get(keyId: string): Promise<ApiKey> {
    const url = `${this._base}${PREFIX}/${encodeURIComponent(keyId)}`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    return ApiKeysClient._unwrap<ApiKey>(resp.json());
  }

  /** Revoke (delete) an API key. */
  async revoke(keyId: string): Promise<void> {
    const url = `${this._base}${PREFIX}/${encodeURIComponent(keyId)}`;
    const resp = await httpRequest(url, { method: "DELETE", headers: this._auth() });
    if (resp.status !== 200 && resp.status !== 204)
      await this._throw(resp.status, resp.text);
  }

  /**
   * Rotate an API key. Returns a row with the new `key` (and typically a
   * new `key_id`). Store the new key immediately — the server cannot
   * return it again.
   */
  async rotate(keyId: string): Promise<ApiKey> {
    const url = `${this._base}${PREFIX}/${encodeURIComponent(keyId)}/rotate`;
    const resp = await httpRequest(url, { method: "POST", headers: this._auth() });
    if (resp.status !== 200 && resp.status !== 201)
      await this._throw(resp.status, resp.text);
    return ApiKeysClient._unwrap<ApiKey>(resp.json());
  }
}
